SimPEG/Directives.py

```diff
@@ -811,3 +832,101 @@ def JtJv(v):
             JtJdiag = JtJdiag / max(JtJdiag)
 
             self.reg.wght = JtJdiag
+
+
+class UpdateSensitivityWeights(InversionDirective):
+    """
+    Directive to take care of re-weighting
+    the non-linear magnetic problems.
+
+    """
+
+    mapping = None
+    JtJdiag = None
+    everyIter = True
+    threshold = 1e-12
+
+    def initialize(self):
+
+        # Calculate and update sensitivity
+        # for optimization and regularization
+        self.update()
+
+    def endIter(self):
+
+        if self.everyIter:
+            # Update inverse problem
+            self.update()
+
+    def update(self):
+
+        # Get sum square of columns of J
+        self.getJtJdiag()
+
+        # Compute normalized weights
+        self.wr = self.getWr()
+
+        # Send a copy of JtJdiag for the preconditioner
+        self.updateOpt()
+
+        # Update the regularization
+        self.updateReg()
+
+    def getJtJdiag(self):
+        """
+            Compute explicitely the main diagonal of JtJ
+            Good for any problem where J is formed explicitely
+        """
+        self.JtJdiag = []
+
+        for prob, survey, dmisfit in zip(self.prob,
```

Comment:
would you mind updating the bracketing / line breaks (similar to above comment)

Reply:
done